services/dedup/badger/badger.go

```diff
@@ -40,45 +40,42 @@ func DefaultPath() string {
 	return fmt.Sprintf(`%v%v`, tmpDirPath, badgerPathName)
 }
 
-func NewBadgerDB(path string) *BadgerDB {
-	dedupWindow := config.GetReloadableDurationVar(3600, time.Second, "Dedup.dedupWindow", "Dedup.dedupWindowInS")
-
-	log := logger.NewLogger().Child("dedup")
+func NewBadgerDB(conf *config.Config, stats stats.Stats, path string) *Dedup {
+	dedupWindow := conf.GetReloadableDurationVar(3600, time.Second, "Dedup.dedupWindow", "Dedup.dedupWindowInS")
+	log := logger.NewLogger().Child("Dedup")
 	badgerOpts := badger.
 		DefaultOptions(path).
 		WithCompression(options.None).
 		WithIndexCacheSize(16 << 20). // 16mb
 		WithNumGoroutines(1).
-		WithNumMemtables(config.GetInt("BadgerDB.numMemtable", 5)).
-		WithValueThreshold(config.GetInt64("BadgerDB.valueThreshold", 1048576)).
+		WithNumMemtables(conf.GetInt("BadgerDB.numMemtable", 5)).
+		WithValueThreshold(conf.GetInt64("BadgerDB.valueThreshold", 1048576)).
 		WithBlockCacheSize(0).
 		WithNumVersionsToKeep(1).
-		WithNumLevelZeroTables(config.GetInt("BadgerDB.numLevelZeroTables", 5)).
-		WithNumLevelZeroTablesStall(config.GetInt("BadgerDB.numLevelZeroTablesStall", 15)).
-		WithSyncWrites(config.GetBool("BadgerDB.syncWrites", false)).
-		WithDetectConflicts(config.GetBool("BadgerDB.detectConflicts", false))
+		WithNumLevelZeroTables(conf.GetInt("BadgerDB.numLevelZeroTables", 5)).
+		WithNumLevelZeroTablesStall(conf.GetInt("BadgerDB.numLevelZeroTablesStall", 15)).
+		WithSyncWrites(conf.GetBool("BadgerDB.syncWrites", false)).
+		WithDetectConflicts(conf.GetBool("BadgerDB.detectConflicts", false))
 
 	db := &BadgerDB{
-		stats:  stats.Default,
+		stats:  stats,
 		logger: loggerForBadger{log},
 		path:   path,
 		gcDone: make(chan struct{}),
 		close:  make(chan struct{}),
 		window: dedupWindow,
 		opts:   badgerOpts,
 	}
-	return db
+	return &Dedup{
+		badgerDB: db,
+		cache:    make(map[string]int64),
+	}
 }
 
 func (d *BadgerDB) Get(key string) (int64, bool, error) {
 	var payloadSize int64
 	var found bool
-	var err error
-	err = d.init()
-	if err != nil {
-		return 0, false, err
-	}
-	err = d.badgerDB.View(func(txn *badger.Txn) error {
+	err := d.badgerDB.View(func(txn *badger.Txn) error {
 		item, err := txn.Get([]byte(key))
 		if err != nil {
 			return err
```

Comment:
[question on line 83] Probably not related to this PR but it seems weird so I'll ask.

Why do we do this?

```go
			payloadSize, _ = strconv.ParseInt(string(itemValue), 10, 64)
			found = true
```

Wouldn't it make more sense to do this instead?

```go
			payloadSize, err = strconv.ParseInt(string(itemValue), 10, 64)
			if err == nil {
				found = true
			}
```

I'm saying this, because if we can get the value from the database but it cannot be parsed then we wrote something invalid. We might as well say we couldn't find it? :thinking: WDYT?

Reply:
I think this was done because we care more whether the key was present or not and not the value for the key